src/js_native_api_v8.cc

```diff
@@ -681,6 +681,155 @@ inline napi_status Wrap(napi_env env,
   return GET_RETURN_STATUS(env);
 }
 
+napi_status DefineStaticProperties(napi_env env,
+                                   napi_value dest,
+                                   size_t static_property_count,
+                                   size_t property_count,
+                                   const napi_property_descriptor* properties) {
+  if (static_property_count > 0) {
+    std::vector<napi_property_descriptor> static_descriptors;
+    static_descriptors.reserve(static_property_count);
+
+    for (size_t i = 0; i < property_count; i++) {
+      const napi_property_descriptor* p = properties + i;
+      if ((p->attributes & napi_static) != 0) {
+        static_descriptors.push_back(*p);
+      }
+    }
+
+    STATUS_CALL(napi_define_properties(env,
+                                       dest,
+                                       static_descriptors.size(),
+                                       static_descriptors.data()));
+  }
+  return napi_clear_last_error(env);
+}
+
+class JSClassGen {
+ public:
+  JSClassGen(const char* utf8name, size_t length) {
+    if (length == NAPI_AUTO_LENGTH)
+      classname = std::string(utf8name);
+    else
+      classname = std::string(utf8name, length);
+    args.resize(3);
+  }
+
+  inline void
+  PostProcess(const std::string& key, const napi_property_descriptor* prop) {
+    if (prop->attributes == napi_default && prop->value == nullptr) return;
+
+    post_process += "    " + key + ": {\n";
```

Comment:
What if `key` contains characters that are not valid identifiers?

Reply:
`napi_run_script` will return non-`napi_ok` if compiling the script fails. The caller can deal with that.

Reply:
Yes, but that’s still a bug?

Reply:
You're right. We should never concat user-supplied strings into the JS snippet directly, because they can contain weird characters. I changed the code so that key is always `[propn]`, and, if the prop descriptor provides a `utf8name` instead of a `name`, it `napi_create_string_utf8()` first and then passes that as a parameter into the resulting function. So, the JS snippet never actually contains literal strings supplied by the user.